sdk/eventhub/event-hubs/src/eventDataBatch.ts

```diff
@@ -258,42 +288,50 @@ export class EventDataBatchImpl implements EventDataBatch {
     const previouslyInstrumented = Boolean(
       eventData.properties && eventData.properties[TRACEPARENT_PROPERTY]
     );
+    let spanContext: SpanContext | undefined;
     if (!previouslyInstrumented) {
       const messageSpan = createMessageSpan(options.parentSpan);
       eventData = instrumentEventData(eventData, messageSpan);
-      this._spanContexts.push(messageSpan.context());
+      spanContext = messageSpan.context();
       messageSpan.end();
     }
+
     // Convert EventData to AmqpMessage.
     const amqpMessage = toAmqpMessage(eventData, this._partitionKey);
     amqpMessage.body = this._context.dataTransformer.encode(eventData.body);
+    const encodedMessage = message.encode(amqpMessage);
 
-    // Encode every amqp message and then convert every encoded message to amqp data section
-    this._encodedMessages.push(message.encode(amqpMessage));
-
-    const batchMessage: AmqpMessage = {
-      body: message.data_sections(this._encodedMessages)
-    };
+    let currentSize = this._sizeInBytes;
+    // The first time an event is added, we need to calculate
+    // the overhead of creating an AMQP batch, including the
+    // message_annotations that are taken from the 1st message.
+    if (this.count === 0) {
+      if (amqpMessage.message_annotations) {
+        this._batchAnnotations = amqpMessage.message_annotations;
+      }
 
-    if (amqpMessage.message_annotations) {
-      batchMessage.message_annotations = amqpMessage.message_annotations;
+      // Figure out the overhead of creating a batch by generating an empty batch
+      // with the expected batch annotations.
+      currentSize += this._generateBatch([], this._batchAnnotations).length;
     }
 
-    const encodedBatchMessage = message.encode(batchMessage);
-    const currentSize = encodedBatchMessage.length;
+    const messageSize = encodedMessage.length;
+    const messageOverhead =
+      messageSize <= smallMessageMaxBytes ? smallMessageOverhead : largeMessageOverhead;
```

Comment:
I can't respect you anymore. Where's the mysterious array with just two constants in it!?